erts/emulator/beam/erl_gc.c

```diff
@@ -1004,131 +1000,85 @@ garbage_collect_hibernate(Process* p, int check_long_gc)
+     * stack.
+     *
+     * IMPORTANT: We have garbage collected to a temporary heap and then copy
+     * the result to a newly allocated heap of exact size.
+     *
+     * !! This is intentional !! Garbage collecting as usual and then shrinking
+     * the heap by reallocating it caused serious fragmentation problems when
+     * large amounts of processes were hibernated. */
+    final_heap = ERTS_HEAP_ALLOC(ERTS_ALC_T_HEAP, sizeof(Eterm) * final_size);
+    sys_memcpy(final_heap, collection_heap, heap_size * sizeof(Eterm));
+    sys_memcpy(&final_heap[final_size - stack_size],
+               p->stop,
+               stack_size * sizeof(Eterm));
 
-    FLAGS(p) &= ~F_FORCE_GC;
-    p->live_hf_end = ERTS_INVALID_HFRAG_PTR;
+    stack_offset = final_heap - p->stop;
+    heap_offset = final_heap - collection_heap;
 
-    /*
-     * Move the heap to its final destination.
-     *
-     * IMPORTANT: We have garbage collected to a temporary heap and
-     * then copy the result to a newly allocated heap of exact size.
-     * This is intentional and important! Garbage collecting as usual
-     * and then shrinking the heap by reallocating it caused serious
-     * fragmentation problems when large amounts of processes were
-     * hibernated.
-     */
+    area = (char *) collection_heap;
+    area_sz = heap_size * sizeof(Eterm);
 
-    ASSERT(actual_size < p->heap_sz);
+    erts_deallocate_young_generation(p);
+    erts_free(ERTS_ALC_T_TMP_HEAP, collection_heap);
 
-    heap = ERTS_HEAP_ALLOC(ERTS_ALC_T_HEAP, sizeof(Eterm)*heap_size);
-    sys_memcpy((void *) heap, (void *) p->heap, actual_size*sizeof(Eterm));
-    ERTS_HEAP_FREE(ERTS_ALC_T_TMP_HEAP, p->heap, p->heap_sz*sizeof(Eterm));
+    p->heap = final_heap;
+    p->heap_sz = final_size;
+    p->high_water = &final_heap[heap_size];
+    p->htop = &final_heap[heap_size];
+    p->hend = &final_heap[final_size];
+    p->stop = &p->heap[final_size - stack_size];
 
-    p->hend = heap + heap_size;
-    p->stop = p->hend - CP_SIZE;
- 
-    switch (erts_frame_layout) {
-    case ERTS_FRAME_LAYOUT_RA:
-        p->stop[0] = make_cp(beam_normal_exit);
-        break;
-    case ERTS_FRAME_LAYOUT_FP_RA:
-        p->stop[0] = make_cp(NULL);
-        p->stop[1] = make_cp(beam_normal_exit);
-        FRAME_POINTER(p) = &p->stop[0];
-        break;
-    }
-
-    offs = heap - p->heap;
-    area = (char *) p->heap;
-    area_sz = ((char *) p->htop) - area;
-    offset_heap(heap, actual_size, offs, area, area_sz);
-    p->high_water = heap + (p->high_water - p->heap);
-    offset_rootset(p, offs, 0, area, area_sz, p->arg_reg, p->arity);
-    p->htop = heap + actual_size;
-    p->heap = heap;
-    p->heap_sz = heap_size;
+    FLAGS(p) &= ~F_FORCE_GC;
+    p->live_hf_end = ERTS_INVALID_HFRAG_PTR;
 
+    offset_heap(final_heap, heap_size, heap_offset, area, area_sz);
+    offset_rootset(p, heap_offset, stack_offset, area, area_sz,
+                   p->arg_reg, p->arity);
```

Comment:
```suggestion
                   p->arg_reg, p->arity);
                   if (erts_frame_layout == ERTS_FRAME_LAYOUT_FP_RA)
                       FRAME_POINTER(p) = &p->stop[0];
```

You need to update the frame pointer as it has been moved by the GC.